devel/query.lua

```diff
@@ -1,254 +1,507 @@
     end
+end
 
-    --Set Key [boolean parsing]
-    if args.setkey == "true" then
-        newkeyvalue=true
-    elseif args.setkey == "false" then
-        newkeyvalue=false
+local bRunOnce={}
+function runOnce(caller)
+    if bRunOnce[caller] == true then
+        return false
     end
+    bRunOnce[caller] = true
+    return true
 end
 
-function debugf(level,...)
-    if args.debug and level <= tonumber(args.debug) then
-        str=string.format(" #  %s",select(1, ...))
-        for i = 2, select('#', ...) do
-            str=string.format("%s\t%s",str,select(i, ...))
+function toType(str)
+    if str ~= nil then
+        if str == "true" then
+            return true
+        elseif str == "false" then
+            return false
+        elseif str == "nil" then
+            return nil
+        elseif tonumber(str) then
+            return tonumber(str)
+        else
+            return tostring(str)
+        end
+    end
+    return nil
+end
+
+--Section: filters
+function is_searchable(field, value)
+    if not is_blacklisted(field, value) and not df.isnull(value) then
+        debugf(3,string.format("is_searchable( %s ): type: %s, length: %s, count: %s", value,type(value),getTableLength(value), countTableLength(value)))
+        if not isEmpty(value) then
+            if getmetatable(value) then
+                if value._kind == "primitive" then
+                    return false
+                elseif value._kind == "struct" then
+                    if args.safer then
+                        return false
+                    else
+                        return true
+                    end
+                end
+                debugf(3,string.format("_kind: %s, _type: %s", value._kind, value._type))
+            end
+            for _,_ in safe_pairs(value) do
+                return true
+            end
+        end
+    end
+    return false
+end
+
+function is_match(path, field, value)
+    if not args.search or string.find(tostring(field),args.search) or string.find(path,args.search) then
+        if not args.findvalue or (not type(value) == "string" and value == find_value) or string.find(value,find_value) then
+            return true
         end
-        print(str)
     end
+    return false
+end
+
+function is_looping(path, field)
+    return not args.dumb and string.find(path, tostring(field))
 end
 
---thanks goes to lethosor for this
-function safe_pairs(item, keys_only)
+function is_blacklisted(field, t)
+    field = tostring(field)
+    if not args.noblacklist then
+        if string.find(field,"script") then
+            return true
+        elseif string.find(field,"saves") then
+            return true
+        elseif string.find(field,"movie") then
+            return true
+        elseif string.find(field,"font") then
+            return true
+        elseif string.find(field,"texpos") then
+            return true
+        end
```

Comment:
Could this function also take the type of `t` into account, assuming it's the table that has these fields?

Also, can the `string.find()` calls be changed to string equality checks? (Not sure which of these are exact field names and which are prefixes/suffixes; if any are prefixes/suffixes and not substrings that could occur anywhere, you could use `string.startswith()` and `string.endswith()` respectively)

Reply:
I don't think I follow. You want to blacklist an entire table? Not data fields? Cause I don't think any of the matching data are tables. Maybe you could elaborate?

Reply:
No, for example, I want to *only* exclude fields with "font" in their name if they are actually the fields that we want to exclude. If `t` is the object containing `field`, then checking the type of `t` would be a way to ensure this.

Reply:
As for my second point about `string.find()`: If you only want to exclude fields named exactly "movie", it would be better to check for `field == "movie"` rather than `string.find(field, "movie")`. If this is possible to change, it would _limit_ the possible false-positives excluded by this function - it would only exclude fields named exactly "movie" in some other type, rather than any field containing "movie" in its name. Since this isn't a complete fix, I would still prefer to have this check the type of the containing object.

Reply:
I tried to re-find the problem fields and could not.. at first. So I played around and realized that the max table length defaults were keeping the problem fields from flooding the console. So I'll just remove this feature altogether as it seems a blacklist is no longer needed..

I could convert it to an opt-in option which takes a list of names to exclude. What do you think?

Reply:
If it's no longer necessary, it's fine with me to remove it. I was actually recently thinking that a list of fields to _include_ would be useful, but I suppose the inverse would be useful as well.

Reply:
Well I think it'd be too complicated to do with fields via the `getfield` option.. but that could work pretty easily with the search option. I'll get about implementing this stuff once it's not 30-45 Celsius around here. Barely have the energy to sit

Reply:
No worries. I can remove this, maybe touch up the docs, and merge if you're ok with it as-is.

Reply:
Unless I missed something this should be pretty decent now. I may, or may not make a secondary help doc sans RST formatting. The blacklist option has been removed.